File: golib/structures.go
// Package golib содержит общие структуры данных, которые часто встречаются
// в задачах LeetCode (связные списки, деревья), и хелперы для их построения
// из срезов — удобно использовать в тестах.
package golib

// ListNode — узел односвязного списка (определение LeetCode).
type ListNode struct {
	Val  int
	Next *ListNode
}

// BuildList строит связный список из среза значений и возвращает его голову.
// Для пустого среза возвращает nil.
func BuildList(values []int) *ListNode {
	dummy := &ListNode{}
	cur := dummy
	for _, v := range values {
		cur.Next = &ListNode{Val: v}
		cur = cur.Next
	}
	return dummy.Next
}

// ListToSlice разворачивает связный список обратно в срез значений.
func ListToSlice(head *ListNode) []int {
	out := []int{}
	for n := head; n != nil; n = n.Next {
		out = append(out, n.Val)
	}
	return out
}

// TreeNode — узел бинарного дерева (определение LeetCode).
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// BuildTree строит бинарное дерево из level-order представления, где nil
// обозначает отсутствующий узел (как в формате ввода LeetCode).
// Пример: []*int{ptr(1), nil, ptr(2), ptr(3)}.
func BuildTree(values []*int) *TreeNode {
	if len(values) == 0 || values[0] == nil {
		return nil
	}
	root := &TreeNode{Val: *values[0]}
	queue := []*TreeNode{root}
	i := 1
	for len(queue) > 0 && i < len(values) {
		node := queue[0]
		queue = queue[1:]
		if i < len(values) && values[i] != nil {
			node.Left = &TreeNode{Val: *values[i]}
			queue = append(queue, node.Left)
		}
		i++
		if i < len(values) && values[i] != nil {
			node.Right = &TreeNode{Val: *values[i]}
			queue = append(queue, node.Right)
		}
		i++
	}
	return root
}

File: lib/structures.ts
// Общие структуры данных, часто встречающиеся в задачах LeetCode
// (связные списки, деревья), и хелперы для их построения из массивов.

/** Узел односвязного списка (определение LeetCode). */
export class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val = 0, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

/** Строит связный список из массива значений; для пустого массива — null. */
export function buildList(values: number[]): ListNode | null {
  const dummy = new ListNode();
  let cur = dummy;
  for (const v of values) {
    cur.next = new ListNode(v);
    cur = cur.next;
  }
  return dummy.next;
}

/** Разворачивает связный список обратно в массив значений. */
export function listToArray(head: ListNode | null): number[] {
  const out: number[] = [];
  for (let n = head; n !== null; n = n.next) out.push(n.val);
  return out;
}

/** Узел бинарного дерева (определение LeetCode). */
export class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(
    val = 0,
    left: TreeNode | null = null,
    right: TreeNode | null = null,
  ) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

/**
 * Строит бинарное дерево из level-order представления, где null обозначает
 * отсутствующий узел (как в формате ввода LeetCode).
 * Пример: [1, null, 2, 3].
 */
export function buildTree(values: (number | null)[]): TreeNode | null {
  if (values.length === 0 || values[0] === null) return null;
  const root = new TreeNode(values[0]);
  const queue: TreeNode[] = [root];
  let i = 1;
  while (queue.length > 0 && i < values.length) {
    const node = queue.shift()!;
    if (i < values.length && values[i] !== null) {
      node.left = new TreeNode(values[i] as number);
      queue.push(node.left);
    }
    i++;
    if (i < values.length && values[i] !== null) {
      node.right = new TreeNode(values[i] as number);
      queue.push(node.right);
    }
    i++;
  }
  return root;
}
